src/main/java/org/apache/sysml/runtime/instructions/gpu/context/GPUMemoryManager.java

```diff
@@ -0,0 +1,561 @@
+	 * If it has zero locks on it, it can be freed, but it is preferrable to keep it around
+	 * so that an extraneous host to dev transfer can be avoided
+	 */
+	private ArrayList<GPUObject> allocatedGPUObjects = new ArrayList<>();
+	
+	/**
+	 * To record size of allocated blocks
+	 */
+	private HashMap<Pointer, Long> allocatedGPUPointers = new HashMap<>();
+	
+	/**
+	 * Adds the GPU object to the memory manager
+	 * 
+	 * @param gpuObj the handle to the GPU object
+	 */
+	public void addGPUObject(GPUObject gpuObj) {
+		allocatedGPUObjects.add(gpuObj);
+	}
+	
+	/**
+	 * Removes the GPU object from the memory manager
+	 * 
+	 * @param gpuObj the handle to the GPU object
+	 */
+	public void removeGPUObject(GPUObject gpuObj) {
+		if(LOG.isDebugEnabled())
+			LOG.debug("Removing the GPU object: " + gpuObj);
+		allocatedGPUObjects.removeIf(a -> a.equals(gpuObj));
+	}
+	
+	
+	/**
+	 * Get size of allocated GPU Pointer
+	 * @param ptr pointer to get size of
+	 * @return either the size or -1 if no such pointer exists
+	 */
+	public long getSizeAllocatedGPUPointer(Pointer ptr) {
+		if(allocatedGPUPointers.containsKey(ptr)) {
+			return allocatedGPUPointers.get(ptr);
+		}
+		return -1;
+	}
+	
+	public GPUMemoryManager(GPUContext gpuCtx) {
+		long free[] = { 0 };
+		long total[] = { 0 };
+		cudaMemGetInfo(free, total);
+		if(free[0] < 0.7*total[0]) {
+			LOG.warn("Potential under-utilization: GPU memory - Total: " + (total[0] * (1e-6)) + " MB, Available: " + (free[0] * (1e-6)) + " MB on " + gpuCtx 
+					+ ". This can happen if there are other processes running on the GPU at the same time.");
+		}
+		else
+			LOG.info("GPU memory - Total: " + (total[0] * (1e-6)) + " MB, Available: " + (free[0] * (1e-6)) + " MB on " + gpuCtx);
+		if (GPUContextPool.initialGPUMemBudget() > OptimizerUtils.getLocalMemBudget()) {
+			LOG.warn("Potential under-utilization: GPU memory (" + GPUContextPool.initialGPUMemBudget()
+					+ ") > driver memory budget (" + OptimizerUtils.getLocalMemBudget() + "). "
+					+ "Consider increasing the driver memory budget.");
+		}
+	}
+	
+	/**
+	 * Invoke cudaMalloc
+	 * 
+	 * @param A pointer
+	 * @param size size in bytes
+	 * @return allocated pointer
+	 */
+	private Pointer cudaMallocWarnIfFails(Pointer A, long size) {
+		try {
+			cudaMalloc(A, size);
+			allocatedGPUPointers.put(A, size);
+			return A;
+		} catch(jcuda.CudaException e) {
+			LOG.warn("cudaMalloc failed immediately after cudaMemGetInfo reported that memory of size " + size + " is available. "
+					+ "This usually happens if there are external programs trying to grab on to memory in parallel.");
+			return null;
+		}
+	}
+	
+	/**
+	 * Allocate pointer of the given size in bytes.
+	 * 
+	 * @param opcode instruction name
+	 * @param size size in bytes
+	 * @return allocated pointer
+	 * @throws DMLRuntimeException if error
+	 */
+	public Pointer malloc(String opcode, long size) throws DMLRuntimeException {
+		if(size < 0) {
```

Comment:
The refactoring is done to minimize indirections, which made it difficult to catch GPU memory bugs.

Reply:
a refactor would have been fine, but this rewritten method omits some functionality that previously existed and is more monolithic.

Reply:
It might be useful if you can construct an example (let's say by reducing utilization factor if necessary) that reproduces the omitted functionality that either makes this PR slower (or makes it crash) when compared to master.

Reply:
The functionality is what i've mentioned in the next couple of comments I made.